### PR TITLE
Space Optimization Fix

### PR DESCRIPTION
**Pull request by GuyAv46**
TL;DR
This PR adds few missing functions for the optimizations (SIMD 4) and improves the logic of choosing the distance function
This PR also includes a benchmark for the distance functions.

This PR is revisiting the optimized distance functions and improve them

1. Added missing functions on AVX512 and AVX implementations
2. improved the choosing logic to take the best function for each dimension. chech `L2_space.cpp` or `IP_space.cpp` for full details
3. reviewed all implementations, aligned operations (all implementations uses shifts on start and not division), made a few improvements on the summing of the results.
4. compared alternative functions by the assembly files for the best flow
5. added more tests for the functions to get a full coverage.
6. added a benchmark for all functions to check future attempts of improvements and confirm the optimization choosing logic

### CODE REVIEW

**Review by alonre24 — COMMENTED** [review]
Looks good! small comments

File: tests/benchmark/CMakeLists.txt [hunk 1]
```diff
@@ -8,9 +8,32 @@ set(CMAKE_CXX_STANDARD 14)
 
 set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -O3 -g")
 
+if(CMAKE_HOST_SYSTEM_PROCESSOR MATCHES "(x86_64)|(AMD64|amd64)|(^i.86$)")
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply tests/benchmark/CMakeLists.txt:11]: I prefer this way because of the way we declare tests in google benchmarks

**Review by alonre24 — APPROVED** [review]
Great job!

File: tests/unit/test_spaces.cpp [hunk 2]
```diff
@@ -143,16 +142,90 @@ TEST_F(SpacesTest, ip_16) {
...
+    case ARCH_OPT_AVX:
+        ASSERT_EQ(baseline, L2SqrSIMD4Ext_AVX(v, v, &dim));
+        optimization = ARCH_OPT_SSE;
+    case ARCH_OPT_SSE:
+        ASSERT_EQ(baseline, L2SqrSIMD4Ext_SSE(v, v, &dim));
+        break;
+    default:
+        ASSERT_TRUE(false);
+    }
+}
```
Comment (DvirDukhan) [inline tests/unit/test_spaces.cpp:228]: please make sure all the functions are tested here

File: src/VecSim/spaces/IP/IP_AVX512.cpp [hunk 3]
```diff
@@ -24,20 +23,63 @@ float InnerProductSIMD16Ext_AVX512_impl(const void *pVect1v, const void *pVect2v
...
+        __m512 v1 = _mm512_loadu_ps(pVect1);
+        pVect1 += 16;
+        __m512 v2 = _mm512_loadu_ps(pVect2);
+        pVect2 += 16;
+        sum512 = _mm512_add_ps(sum512, _mm512_mul_ps(v1, v2));
+        // sum512 = _mm512_fmadd_ps(v1, v2, sum512);
+    }
+
+    __m128 v1, v2;
+    __m128 sum_prod = _mm512_extractf32x4_ps(sum512, 0) + _mm512_extractf32x4_ps(sum512, 1) +
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/VecSim/spaces/IP/IP_AVX512.cpp:62]: not from __m512 to __m128, and this line is pretty efficient

File: src/VecSim/spaces/IP/IP_AVX512.cpp [hunk 4]
```diff
@@ -24,20 +23,63 @@ float InnerProductSIMD16Ext_AVX512_impl(const void *pVect1v, const void *pVect2v
...
+    while (pVect1 < pEnd2) {
+        v1 = _mm_loadu_ps(pVect1);
+        pVect1 += 4;
+        v2 = _mm_loadu_ps(pVect2);
+        pVect2 += 4;
+        sum_prod = _mm_add_ps(sum_prod, _mm_mul_ps(v1, v2));
+    }
+
+    _mm_store_ps(TmpRes, sum_prod);
+    float sum = TmpRes[0] + TmpRes[1] + TmpRes[2] + TmpRes[3];
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/VecSim/spaces/IP/IP_AVX512.cpp:74]: not supported on older compilers, and it wont save to much time (one less command)

File: tests/unit/test_spaces.cpp (hunk 2 above)
Comment (GuyAv46) [reply tests/unit/test_spaces.cpp:228]: All are covered :)

File: src/VecSim/spaces/IP/IP_AVX512.cpp [hunk 5]
```diff
@@ -24,20 +23,63 @@ float InnerProductSIMD16Ext_AVX512_impl(const void *pVect1v, const void *pVect2v
         __m512 v2 = _mm512_loadu_ps(pVect2);
         pVect2 += 16;
         sum512 = _mm512_add_ps(sum512, _mm512_mul_ps(v1, v2));
+        // sum512 = _mm512_fmadd_ps(v1, v2, sum512);
```
Comment (alonre24) [inline src/VecSim/spaces/IP/IP_AVX512.cpp:26]: remove?

File: src/VecSim/spaces/IP_space.cpp [hunk 6]
```diff
@@ -25,7 +36,7 @@ InnerProductSpace::InnerProductSpace(size_t dim, std::shared_ptr<VecSimAllocator
             fstdistfunc_ = InnerProductSIMD16Ext_AVX;
         } else if (dim % 4 == 0) {
             fstdistfunc_ = InnerProductSIMD4Ext_AVX;
-        } else if (dim > 16) {
+        } else if (dim > 16 && dim % 16 < 4) {
```
Comment (alonre24) [inline src/VecSim/spaces/IP_space.cpp:39]: Why lines 39 and 52 are not covered?

File: tests/benchmark/bm_spaces.cpp [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+
+class BM_VecSimSpaces : public benchmark::Fixture {
+protected:
+    std::mt19937 rng;
+    size_t dim;
+    float *v1, *v2;
+    Arch_Optimization opt;
+
+#define ASSERT_OPT(OPT)                                                                            \
+    if (1) {                                                                                       \
```
Comment (alonre24) [inline tests/benchmark/bm_spaces.cpp:20]: Why `if(1)` ?

File: src/VecSim/spaces/IP/IP_AVX512.cpp (hunk 5 above)
Comment (GuyAv46) [reply src/VecSim/spaces/IP/IP_AVX512.cpp:26]: I left it as a reminder that its exists, maybe it will be useful someday.
The compiler should be able to compile both lines into a single `fmadd_ps` command for the CPU

File: src/VecSim/spaces/IP_space.cpp (hunk 6 above)
Comment (GuyAv46) [reply src/VecSim/spaces/IP_space.cpp:39]: making a test for every available optimization is complicated. you have to run the test on different machines to go through all lines here. we test only the functions themself and not the logic.
the only lines that are covered are the lines the coverage test machine is directed to through our other tests

File: tests/benchmark/bm_spaces.cpp (hunk 7 above)
Comment (GuyAv46) [reply tests/benchmark/bm_spaces.cpp]: `if(1) {<code>} else` and `do {<code>} while(0)` are two common ways to encapsulate a macro and make it that it can be treated as a function call, with adding `;` at the end without breaking the code inside the macro

**Review by alonre24 — APPROVED** [review]

**Review by DvirDukhan — COMMENTED** [review]
small comment
also, **please elaborate** on all the modifications done in this PR in the PR description

File: tests/unit/test_spaces.cpp [hunk 8]
```diff
@@ -143,16 +142,90 @@ TEST_F(SpacesTest, ip_16) {
...
+        optimization = ARCH_OPT_SSE;
+    case ARCH_OPT_SSE:
+        ASSERT_EQ(baseline, L2SqrSIMD16Ext_SSE(v, v, &dim));
+        break;
+    default:
+        ASSERT_TRUE(false);
+    }
+}
+
+// This test will trigger the function for dimension % 16 == 0 and dimension % 4 == 0, for each
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply tests/unit/test_spaces.cpp:179]: it's been a few years since high school..

**Review by DvirDukhan — APPROVED** [review]
